File: ai-interview-platform-merged/backend/utils/activityLog.js
const pool = require('../config/db');

/**
 * Append-only, best-effort activity log for Module 1's admin "Monitor
 * system activities". Never throws — a logging failure should never
 * break the actual request it's describing.
 *
 * @param {Object} params
 * @param {number|null} params.userId - actor's user id, or null for system/anonymous events
 * @param {string|null} params.role - actor's role at the time of the action
 * @param {string} params.action - short machine-readable action name, e.g. "user_registered"
 * @param {string} [params.details] - human-readable one-liner shown in the admin UI
 */
async function logActivity({ userId = null, role = null, action, details = '' }) {
  try {
    await pool.query(
      'INSERT INTO activity_log (actor_user_id, actor_role, action, details) VALUES ($1, $2, $3, $4)',
      [userId, role, action, details]
    );
  } catch (err) {
    console.error('Activity log write failed (non-fatal):', err.message);
  }
}

module.exports = { logActivity };

File: ai-interview-platform-merged/backend-python/app/activity_log.py
"""
Module 1 — Admin "Monitor system activities".

Writes to the SAME `activity_log` table the Node service writes to (see
backend/utils/activityLog.js) for registrations/logins/admin actions —
this just covers the events that only happen on this service, mainly
interview completions. Best-effort: a logging failure must never break
the request it's describing.
"""
from sqlalchemy import text
from sqlalchemy.orm import Session


def log_activity(db: Session, user_id: int, role: str, action: str, details: str = "") -> None:
    try:
        db.execute(
            text(
                "INSERT INTO activity_log (actor_user_id, actor_role, action, details) "
                "VALUES (:user_id, :role, :action, :details)"
            ),
            {"user_id": user_id, "role": role, "action": action, "details": details},
        )
        db.commit()
    except Exception:
        db.rollback()
